contracts/limitOrder/LimitOrderBook.sol

```diff
@@ -219,4 +228,66 @@ contract LimitOrderBook is
 
         return signer;
     }
+
+    //
+    // INTERNAL VIEW
+    //
+
+    function _formatDecimals(
+        uint256 price,
+        uint8 fromDecimals,
+        uint8 toDecimals
+    ) internal pure returns (uint256) {
+        // LOB_ID: Invalid Decimals
+        require(fromDecimals <= toDecimals, "LOB_ID");
+
+        return price.mul(10**(toDecimals.sub(fromDecimals)));
+    }
+
+    function _getPriceByRoundId(address baseToken, uint80 roundId) internal view returns (uint256) {
+        ChainlinkPriceFeed chainlinkPriceFeed = ChainlinkPriceFeed(IBaseToken(baseToken).getPriceFeed());
+        (uint256 price, ) = chainlinkPriceFeed.getRoundData(roundId);
+        return _formatDecimals(price, chainlinkPriceFeed.decimals(), 18);
+    }
+
+    function _verifyTriggerPrice(LimitOrder memory order, uint80 roundIdWhenTriggered) internal view {
+        // NOTE: Chainlink proxy's roundId is always increased
+        // https://docs.chain.link/docs/historical-price-data/
+
+        // LOB_IRI: Invalid RoundId
+        require(order.roundIdWhenCreated > 0 && roundIdWhenTriggered > order.roundIdWhenCreated, "LOB_IRI");
+
+        // LOB_ITP: Invalid Trigger Price
+        require(order.triggerPrice > 0, "LOB_ITP");
+
+        // NOTE: we can only support stop/take-profit limit order for markets that use ChainlinkPriceFeed
+        uint256 triggeredPrice = _getPriceByRoundId(order.baseToken, roundIdWhenTriggered);
+
+        // we need to make sure the price has reached trigger price.
+        // however, we can only know whether index price has reached trigger price,
+        // we didn't know whether market price has reached trigger price
+
+        // rules of advanced order types
+        // https://help.ftx.com/hc/en-us/articles/360031896592-Advanced-Order-Types
+        if (order.orderType == ILimitOrderBook.OrderType.StopLimitOrder) {
```

Comment:
I guess we need only two cases? cuz bot runners can tell the order type from the order itself?

Reply:
Maybe even one... hmm... I have no idea and am just asking for opinions!

Reply:
Since there has to be 4 `require` statements, so what's the advantage to only have 1 or 2 error messages?

I consider this is similar to https://github.com/perpetual-protocol/perp-lushan/blob/main/contracts/ClearingHouse.sol#L1069

Reply:
Good question! 
I was thinking that since `fillLimitOrder` is most likely to be called by bots only, we won't be showing these error codes on FE and thus we don't need these 4 variances to let users know what's the exact mistake they make (on FE).

Hence, if the main audience of these error codes is code-writers, perhaps one error code telling something like `trigger price not satisfied` is sufficient... but I don't know; if you're the bot writer, will you need four of them or is one sufficient?

Reply:
Conclusion: better precision!